### ISSUE TITLE
Increase hard limit of 12 waypoints in 'Tap to Route'?

### ISSUE THREAD

**Issue by oliverlockwood** [cyclestreets/android#186]
The Tap to Route with waypoints feature is a great tool for planning a custom route rather than an A-to-B one.

A few things I've noticed:
1.  There's a hard maximum of 12 waypoints, as hardcoded [here](https://github.com/cyclestreets/android/blob/master/libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/TapToRouteOverlay.java#L610).  The [page documenting the underlying API](https://www.cyclestreets.net/api/v2/journey.plan/) says "Information on the new V2 API is only available to Administrators at present."  The old [V1 API documentation](https://www.cyclestreets.net/api/v1/journey/) says "There can be at most 12 waypoints".  **I'd love to know if there's a reason for this, or whether there's any scope for this being relaxed, perhaps as part of a migration to the V2 API (which I could create a PR for if I had access to the documentation!)**
2.  ~~While you're on waypoints 1-11, you can press the "back" button on the UI to remove the last waypoint if you placed it in the wrong place.  However, once you've placed waypoint 12, the back button stops responding and you're stuck.  Which is really annoying if you've painstakingly created a route only to mis-click on the last waypoint.  This is just a bug and needs fixing.~~ This has been addressed by #226.

**Comment by oliverlockwood** [comment]
@jezhiggins @mvl22 any thoughts on item (1) above?

I'll dig into investigating item (2) when I have the chance.

**Comment by mvl22** [comment]
@si-the-pie should be able to clarify why there's a limit of 12 points.

There is still no date for the V2 API. However, if you switch away from the V1 XML to the equivalent V1 JSON output that will presumably be most of the work done, and should enable you at last to rip out the last of the XML parsing stuff.

**Comment by oliverlockwood** [comment]
@si-the-pie - can you please share any clarification on why the 12 point limit?  I agree with @mvl22 that it seems restrictive - which is in fact how I came across this problem.

Clearly an app developer could knit together multiple 12-point routes but it'd be a bit of a pain, and if the restriction isn't necessary it would keep things simpler.

**Comment by oliverlockwood** [comment]
@si-the-pie @mvl22 a year on, would you mind answering my query about the hard-limit of 12 waypoints, please?

**Comment by mvl22** [comment]
The reason this is there is because there was a particular API client implementation that erroneously kept sending past parts of the planned route as the user went along. So a route that starts A-B-C, became A-B-C-D (at the point of reaching B), A-B-C-D-E (at C), etc, with the result that 50+ waypoints are sent, which becomes more computationally expensive.

We want to get rid of this limit, and apply a deprecation scenario to this particular client.

I think it's worth having a limit (the technical limit at present I think would be 256), but not 12. Sounds like this is a priority which we should expedite?

**Comment by oliverlockwood** [comment]
Thanks for the feedback. It sounds like something reasonably easy to control on the server side for a particular client by the API key, so I'd be in favour of it being changed soon if that's reasonably possible.

I think a hard limit does make sense for computational reasons. From personal experience using the app I've tended to want to go a bit over 12 but not massively so. I guess the setting off the new limit would be a balancing act of usefulness against computational cost; I imagine there's an elbow in the graph somewhere?

If I were to pick a number out of thin air I'd say 30, to give some headroom, but even raising it to 20 or 25 would be a significant improvement.

**Comment by si-the-pie** [comment]
The limit on the number of waypoints imposed by the journey planning API has been changed to 30 from 12.

Waypoint stages are queued serially to the routing engine by a php layer, so the number of waypoints sent is effectively a multiplier on the overall response time to generate a route. That was the main reason for keeping it to a low value. Another reason is that the current user interfaces do not provide any control over editing the intermediate waypoints.

I've also set the same limit in the [mobileweb interface](https://m.cyclestreets.net/journey/) (do a forced reload to avoid cached javascript).
